internal/topology/variables/cluster_variable_validation_test.go

```diff
@@ -1899,6 +1899,298 @@ func Test_ValidateClusterVariable(t *testing.T) {
+					"spec.topology.variables[test].value.propertyA",
+				),
+			},
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name:     "test",
+				Required: true,
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "object",
+						AnyOf: []clusterv1.JSONSchemaProps{{
+							Required: []string{"propertyA"},
+						}, {
+							Required: []string{"propertyB"},
+						}},
+						Properties: map[string]clusterv1.JSONSchemaProps{
+							"propertyA": {
+								Type: "boolean",
+							},
+							"propertyB": {
+								Type: "boolean",
+							},
+						},
+					},
+				},
+			},
+			clusterVariable: &clusterv1.ClusterVariable{
+				Name: "test",
+				Value: apiextensionsv1.JSON{
+					Raw: []byte(`{"propertyA":"invalid value"}`),
+				},
+			},
+		},
+		{
+			name: "Valid object with anyOf and not schema",
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name:     "test",
+				Required: true,
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "object",
+						AnyOf: []clusterv1.JSONSchemaProps{{
+							Required: []string{"propertyA"},
+						}, {
+							Required: []string{"propertyB"},
+						}, {
+							Required: []string{"propertyC"},
+						}},
+						Not: &clusterv1.JSONSchemaProps{
+							Required: []string{"propertyA", "propertyB"},
+						},
+						Properties: map[string]clusterv1.JSONSchemaProps{
+							"propertyA": {
+								Type: "boolean",
+							},
+							"propertyB": {
+								Type: "boolean",
+							},
+							"propertyC": {
+								Type: "boolean",
+							},
+						},
+					},
+				},
+			},
+			clusterVariable: &clusterv1.ClusterVariable{
+				Name: "test",
+				Value: apiextensionsv1.JSON{
+					Raw: []byte(`{"propertyA":true, "propertyC":true}`),
+				},
+			},
+		},
+		{
+			name: "Fails, value does match not schema",
+			wantErrs: []validationMatch{
+				invalid(
+					`Invalid value: "{\"propertyA\":true, \"propertyB\":true}": "" must not validate the schema (not)`,
+					"spec.topology.variables[test].value",
+				),
+			},
+			clusterClassVariable: &clusterv1.ClusterClassVariable{
+				Name:     "test",
+				Required: true,
+				Schema: clusterv1.VariableSchema{
+					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+						Type: "object",
+						AnyOf: []clusterv1.JSONSchemaProps{{
+							Required: []string{"propertyA"},
+						}, {
+							Required: []string{"propertyB"},
+						}},
+						Not: &clusterv1.JSONSchemaProps{
+							Required: []string{"propertyA", "propertyB"},
+						},
+						Properties: map[string]clusterv1.JSONSchemaProps{
+							"propertyA": {
+								Type: "boolean",
+							},
+							"propertyB": {
+								Type: "boolean",
+							},
+						},
+					},
+				},
+			},
+			clusterVariable: &clusterv1.ClusterVariable{
+				Name: "test",
+				Value: apiextensionsv1.JSON{
+					Raw: []byte(`{"propertyA":true, "propertyB":true}`),
+				},
+			},
```

Comment:
Let's add one more for the int-or-string

```go
{
			name: "pass & fail correctly for int-or-string in oneOf/anyOf/allOf schemas",
			clusterClassVariable: &clusterv1.ClusterClassVariable{
				Name: "variableA",
				Schema: clusterv1.VariableSchema{
					OpenAPIV3Schema: clusterv1.JSONSchemaProps{
						Type: "object",
						Properties: map[string]clusterv1.JSONSchemaProps{
							"anyOfExampleField": { // Only valid variant
								Type: "object",
								AnyOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
							"allOfExampleField": {
								Type: "object",
								AllOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
							"oneOfExampleField": {
								Type: "object",
								OneOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
							},
						},
					},
				},
			},
			wantErrs: []validationMatch{
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[allOfExampleField].allOf[0].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[allOfExampleField].allOf[1].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[oneOfExampleField].oneOf[0].type"),
				forbidden("must be empty to be structural", "spec.variables[variableA].schema.openAPIV3Schema.properties[oneOfExampleField].oneOf[1].type"),
			},
		},
```

Reply:
I added a couple of test cases to check that `anyOf` and `oneOf` both successfully validate as per OpenAPI schema semantics, `allOf` fails of course due to not being able to satisfy both types.

Also added test case for invalid typed values for all these schemas.

Reply:
I'm not sure if we should add test cases to the variable validation tests that wouldn't pass the schema validation

> Also those nodes are partially excluded from rule 3 in the sense that the following two patterns are allowed (exactly those, without variations in order to additional fields):
> xref: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#intorstring

If I got this correctly the only case where it's valid to define a type is:
```yaml
anyOf:
  - type: integer
  - type: string
```

All other cases (e.g. allOf / oneOf) will be rejected by structural schema validation. (see "pass & fail correctly for int-or-string in oneOf/anyOf/allOf schemas" in clusterclass_variable_validation_test.go)

I guess technically it's fine to have the test cases here because if you would be able to get this variable schema deployed, the variable validation would work that way.

On the other side it could be confusing to have test cases here that are not possible in reality (some folks are using our unit tests to lookup examples of what works)


What do you think?
(I think at the very least I would like to add comments that these variable schemas are not going to pass variable schema validation)

Reply:
That's a good point. Hmm I'm actually surprised that `anyOf` even passes schema validation without `x-kubernetes-int-or-string: true` 🤔 Also it looks wrong to me to specify `type: object` along with the `anyOf` types (see https://github.com/kubernetes-sigs/cluster-api/pull/10637/files#diff-d6b60f4f64593285db6e325b594ce01480668c685f583a74ee5568d81c37d795R2749-R2756)

Reply:
Damn, I was surprised as well but I missed that I was accidentally setting type object :/

Reply:
One moment

Reply:
Okay so here's an example: (resource.Quantitty has the schema below hard-coded in controller-gen)
```go
	TestQuantity resource.Quantity `json:"testQuantity"`
```


YAML in the CRD

````
                    testQuantity:
                      anyOf:
                      - type: integer
                      - type: string
                      pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
                      x-kubernetes-int-or-string: true
```

Reply:
So the following would work but only if we additionally also have the int-or-string field:
```
							"anyOfExampleField": { // Only valid variant
								AnyOf: []clusterv1.JSONSchemaProps{{
									Type: "integer",
								}, {
									Type: "string",
								}},
								Pattern: `^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$`,
							},
```

Reply:
We have two options:
* Also implement int-or-string in this PR
* leave it out of scope of this PR (I would take the follow-up PR)

Reply:
Added XIntOrString to variable schema types, updated conversion func, and updated test cases. Looking good I think.

I think it's best to remove the invalid test cases from the value validation tests as the schemas are invalid and updated the test case you proposed below to have a valid schema (even if it doesn't affect the test case result, it's good to be accurate for others looking at test cases for valid schemas I think).

Reply:
Sounds good, will review again now!